demos/text_detection_demo/src/text_recognition.cpp

```diff
@@ -53,3 +61,91 @@ std::string CTCGreedyDecoder(const std::vector<float> &data, const std::string&
     }
     return res;
 }
+
+std::string CTCBeamSearchDecoder(const std::vector<float> &data, const std::string& alphabet, char pad_symbol, double *conf, int bandwidth) {
+    std::string res = "";
+
+    const int num_classes = alphabet.length();
+
+    std::vector<beam> curr;
+    std::vector<beam> last;
+
+    beam init("", 1.f, 0.f, 1.f);
+    last.push_back(init);
+
+    for (std::vector<float>::const_iterator it = data.begin(); it != data.end(); it += num_classes) {
+        curr.clear();
+
+        std::vector<float> prob = std::vector<float>(num_classes, 0.f);
+        softmax_layer(it, it + num_classes, prob);
+
+        for (int _candidate = 0; _candidate < static_cast<int>(last.size()); _candidate++) {
+            float _pNB = 0.f;
+            auto __can = last[_candidate];
+            std::string __can_sentance = __can.sentance;
+            if (__can_sentance != "") {
+                int n = static_cast<int>(__can_sentance.back());
+                _pNB = __can.pNB * prob[n];
+            }
+            float _pB = __can.pT * prob[(num_classes - 1)];
+
+            auto check_res = std::find_if(curr.begin(), curr.end(), [__can_sentance](beam const& n) {
+                return n.sentance == __can_sentance;
+            });
+            if (check_res == std::end(curr)) {
+                curr.push_back(beam(__can.sentance, _pB, _pNB, _pB + _pNB));
+            } else {
+                auto __i = std::distance(curr.begin(), check_res);
+                curr[__i].pNB += _pNB;
```

Comment:
I wouldn't add this into code since it's very close to zero, but not really zero.

Reply:
I'm sorry, but it MUST be zero in this line of the code. This condition follows from the consistency of the algorithm.

The reason is as follows:
* when a next symbol is received, the algorithm runs through all candidate elements of the vector `last` and adds new elements to the vector `curr`
    (and vector `curr` is cleared at the begin of each iteration)
* elements are added to the vector `curr` in such a way, that different elements are *always* have different values of the field `sentence`
* the vector `last` is a part of the vector `curr` on the previous iteration. 
    Thus the different elements in the vector `last` are *always* have different values of the field `sentence`
* There are only 2 places in the algorithm where an element may be added into the vector `curr` or an element from vector `curr` is modified: 
   * lines 99-115  -- when the new element has the same field `sentence` as the current candidate https://github.com/opencv/open_model_zoo/blob/c2e691890405f9642713dfe474fe18cdc1434b50/demos/text_detection_demo/src/text_recognition.cpp#L99-L115
   * lines 118-135, when a new char is added to the end of the field `sentence` of the current candidate https://github.com/opencv/open_model_zoo/blob/c2e691890405f9642713dfe474fe18cdc1434b50/demos/text_detection_demo/src/text_recognition.cpp#L117-L136
* At that in lines 118-135 the field `prob_blank` is neither modified nor set to non-zero values
    (the new elements which are added in the part of the code always have the field `prob_blank = 0` )
* So, the only part of the code where the field `prob_blank` may be changed to non-zero value is the part in the lines 99-115
* But in this part the field `sentence` of the new element (in the vector `curr`) is the same as the field `sentence` in the candidate element  (in the vector `last`)
* Since for different elements from the vector `last` the field `sentence` is different, this part of code (lines 99-115) may be fulfilled only once for each sentence
* Thus when the algorithm enters this part of the code, the field `prob_blank` MUST be zero (if the algorithm works properly)

Reply:
One more possible question: why such code is required here, if we can prove (as I did it above) that the field `prob_blank` must be zero here
```cpp
if curr[index].prob_blank != 0.f
    throw std::logic_error("Probability that the last char in CTC-sequence is the special blank char must be zero here");
```

My answer is as follows:  the code will be changed and refactored (it always happens, sooner or later), and it is important that even if the refactoring is made by another person, there are some consistency checks in the algorithm to avoid dummy bugs in refactoring.

Reply:
Please, note that @Wovchena and me added this check in the code and found that it works properly.